Add brake test announcer

hslusarek opened the pull request.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/brake_test_filter.h:
```diff
@@ -0,0 +1,105 @@
+private:
+  void cb(const EventType& evt)
+  {
+	  try
```
Somehow your IDE seems to have wrong space settings...

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/modbus_brake_test_announcer.h:
```diff
@@ -0,0 +1,66 @@
+
+private:
+  //! Filters consecutive messages with the same timestamp.
+  //! Passed messages are redirected to callback "modbus_read_callback()".
```
Not true

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/modbus_msg_brake_test_wrapper.h:
```diff
@@ -0,0 +1,150 @@
+#define MODBUS_MSG_BRAKE_TEST_WRAPPER_H
+
+#include <prbt_hardware_support/ModbusMsgInStamped.h>
+#include <prbt_hardware_support/pss4000_modbus_api_definitions.h>
```
We could also formulate the modbus API hardware agnostic. What do you think?

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/modbus_msg_sto_wrapper_exception.h:
```diff
@@ -22,4 +22,4 @@ namespace prbt_hardware_support
 
 }
 
-#endif // MODBUS_MSG_STO_WRAPPER_EXCPETION_H
```
Please fix in separate PR, also there is a typo...

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/pss4000_modbus_api_definitions.h:
```diff
@@ -0,0 +1,50 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef PSS4000_MODBUS_API_DEFINITIONS_H
```
If we move this whole configuration into the parameter server this file would become unnecessary...

agutenkunst commented on prbt_hardware_support/msg/BrakeTestRequired.msg:
```diff
@@ -0,0 +1,20 @@
+#
```
Why not just use http://docs.ros.org/api/std_msgs/html/msg/Bool.html ?

martiniil replied:
Replaced by the standard type, since we do not see a strong reason.

agutenkunst commented on prbt_hardware_support/src/modbus_brake_test_announcer.cpp:
```diff
@@ -0,0 +1,69 @@
+namespace prbt_hardware_support
+{
+
+static constexpr int DEFAULT_QUEUE_SIZE_MODBUS {1000};
```
Do you really want a queue this long in this case?

agutenkunst commented on prbt_hardware_support/src/modbus_brake_test_announcer.cpp:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  int modbus_api_version = msg.getVersion();
+  if(modbus_api_version != MODBUS_API_VERSION_REQUIRED)
```
This comparision is unsigned int with int

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_brake_test_filter.cpp:
```diff
@@ -0,0 +1,183 @@
+		using ::testing::Throw;
+		using namespace prbt_hardware_support;
+
+		class CallbackReceiver {
```
Spacing!

agutenkunst commented on prbt_hardware_support/src/modbus_brake_test_announcer_node.cpp:
```diff
@@ -0,0 +1,35 @@
+    ros::init(argc, argv, "modbus_brake_test_announcer");
+    ros::NodeHandle nh {"~"};
+
+    prbt_hardware_support::ModbusBrakeTestAnnouncer brake_test_announcer(nh);
```
Maybe here we could open this for extension by using a switch case over a parameter specifying the Subclass to use?

agutenkunst reviewed (changes requested):
Some changes requested. If we configure the protocol details (registers) via yaml there is still some work to do...

martiniil committed:
Announce braketest for SafeBrakeControl (SBC)

A braketest requirement by the safety controller is
announce via ROS topic. The modbus API was extended with
a flag indicating a pending braketest request which is translated
into a boolean ROS message.

martiniil commented on prbt_hardware_support/include/prbt_hardware_support/modbus_api_spec.h:
```diff
@@ -0,0 +1,42 @@
+
+/**
+ * @brief Publishes a message informing about a required brake test.
+ */
```
Please fix.
